src/pymmcore_plus/mda/_engine.py

```diff
@@ -54,11 +69,46 @@ def setup_event(self, event: MDAEvent) -> None:
         self._mmc.waitForSystem()
 
     def exec_event(self, event: MDAEvent) -> Any:
-        """Execute an individual event and return the image data."""
-        # TODO: add non-aquisition event-specific logic here later
+        """Execute an individual event."""
+        action = getattr(event, "action", None)
+
+        # execute hardware autofocus
+        if isinstance(action, HardwareAutofocus) and event.z_pos is not None:
+            # get position index
+            p_idx = event.index.get("p", None)
+            # run autofocus and get the new z position
+            new_z = self._execute_autofocus(action)
+            # get the correction to apply to each z position
+            self._z_correction[p_idx] = new_z - event.z_pos
```

Comment:
```suggestion
            # update the correction for this position
            self._z_correction[p_idx] = self._execute_autofocus(action)
```
i think it would read _slightly_ clearer if `self._execute_autofocus()` returned a delta, ie. it should return the `new_z - event.z_pos` rather than just the `new_z`.  And to do so, it should use `self._mmc.getPosition()` inside of the body of `_execute_autofocus` rather than using `event.z_pos` here (which is correct of course, but only insofar as `setup_event()` *elsewhere* actually set the position to event.z_pos.  So this is just a bit more explicit

Reply:
I used the `event.z_pos` mainly because I thought that in case of a relative z-stack, we are more precise if we use the event info instead of the stage position since we want the slice interval to be precisely what we set in the `MDASequence`. But maybe it really doesn't matter and is the same. Do you mean something like this:

```py
def exec_event(self, event: MDAEvent) -> Any:
        """Execute an individual event."""
        action = getattr(event, "action", None)

        # execute hardware autofocus
        if isinstance(action, HardwareAutofocus) and event.z_pos is not None:
            # get position index
            p_idx = event.index.get("p", None)
            # run autofocus and get the correction to apply to each z position
            self._z_correction[p_idx] = self._execute_autofocus(action)

            return None

        # acquire an image and emit the image data
        self._mmc.snapImage()
        return EventPayload(image=self._mmc.getImage())

    def _execute_autofocus(self, action: HardwareAutofocus) -> float:
        """Perform the hardware autofocus.

        Returns the z correction to apply to each z position.
        """
        self._mmc.setPosition(
            action.autofocus_device_name,
            action.autofocus_motor_offset,
        )
        self._mmc.waitForSystem()
        
        current_z = self._mmc.getZPosition()

        @retry(exceptions=RuntimeError, tries=action.max_retries, logger=logger.warning)
        def _full_focus() -> float:
            self._mmc.fullFocus()
            self._mmc.waitForSystem()
            return self._mmc.getZPosition() - current_z

        try:
            return _full_focus()
        except RuntimeError:
            logger.warning("Warning: Hardware autofocus failed.")
            return self._mmc.getZPosition()
```